wolfcrypt/src/evp.c

```diff
@@ -1908,15 +1922,132 @@ int wolfSSL_EVP_PKEY_sign(WOLFSSL_EVP_PKEY_CTX *ctx, unsigned char *sig,
-        WOLFSSL_MSG("not implemented");
-        FALL_THROUGH;
+#ifndef NO_DSA
+    case EVP_PKEY_DSA:
+        if (wolfSSL_DSA_do_sign(tbs, sig, ctx->pkey->dsa) == WOLFSSL_SUCCESS) {
+            *siglen = wolfSSL_BN_num_bytes(ctx->pkey->dsa->q);
+            return WOLFSSL_SUCCESS;
+        } else {
+            return WOLFSSL_FAILURE;
+        }
+#endif /* NO_DSA */
+
+#ifdef HAVE_ECC
+    case EVP_PKEY_EC: {
+        WOLFSSL_ECDSA_SIG *ecdsaSig = wolfSSL_ECDSA_do_sign(tbs, (int)tbslen, ctx->pkey->ecc);
+        if (ecdsaSig == NULL)
+            return WOLFSSL_FAILURE;
+        *siglen = wolfSSL_i2d_ECDSA_SIG(ecdsaSig, &sig);
+        wolfSSL_ECDSA_SIG_free(ecdsaSig);
+        return WOLFSSL_SUCCESS;
+    }
+#endif /* HAVE_ECC */
+
     default:
         break;
     }
     return WOLFSSL_FAILURE;
 }
 
+/******************************************************************************
+* wolfSSL_EVP_PKEY_verify_init -  initializes a public key algorithm context for
+* a verification operation.
+*
+* RETURNS:
+* returns WOLFSSL_SUCCESS on success, WOLFSSL_FAILURE on failure. In particular
+* a return value of -2 indicates the operation is not supported by the public
+* key algorithm.
+*/
+
+int wolfSSL_EVP_PKEY_verify_init(WOLFSSL_EVP_PKEY_CTX *ctx)
+{
+    WOLFSSL_MSG("wolfSSL_EVP_PKEY_verify_init");
+
+    if (!ctx || !ctx->pkey)
+        return WOLFSSL_FAILURE;
+
+    switch (ctx->pkey->type) {
+#if !defined(NO_RSA) && !defined(HAVE_USER_RSA)
+        case EVP_PKEY_RSA:
+            ctx->op = EVP_PKEY_OP_VERIFY;
+            return WOLFSSL_SUCCESS;
+#endif /* NO_RSA */
+
+#ifndef NO_DSA
+        case EVP_PKEY_DSA:
+            ctx->op = EVP_PKEY_OP_VERIFY;
+            return WOLFSSL_SUCCESS;
+#endif /* NO_DSA */
+
+#ifdef HAVE_ECC
+        case EVP_PKEY_EC:
+            ctx->op = EVP_PKEY_OP_VERIFY;
+            return WOLFSSL_SUCCESS;
+#endif /* HAVE_ECC */
+
+        default:
+            return -2;
+    }
+
+    return WOLFSSL_FAILURE;
+}
+
+/******************************************************************************
+* wolfSSL_EVP_PKEY_verify - verifies a signiture using ctx
+*
+* RETURNS:
+* returns WOLFSSL_SUCCESS on success, WOLFSSL_FAILURE on failure. In particular
+* a return value of -2 indicates the operation is not supported by the public
+* key algorithm.
+*/
+
+int wolfSSL_EVP_PKEY_verify(WOLFSSL_EVP_PKEY_CTX *ctx, const unsigned char *sig,
+                            size_t siglen, const unsigned char *tbs, size_t tbslen)
+{
+    WOLFSSL_MSG("wolfSSL_EVP_PKEY_verify");
+
+    if (!ctx || ctx->op != EVP_PKEY_OP_VERIFY || !ctx->pkey)
+        return WOLFSSL_FAILURE;
+
+    switch (ctx->pkey->type) {
+#if !defined(NO_RSA) && !defined(HAVE_USER_RSA)
+    case EVP_PKEY_RSA: {
+        unsigned char md[WC_MAX_DIGEST_SIZE];
+        int md_len = wolfSSL_RSA_public_decrypt(
+            (int)siglen, sig, md, ctx->pkey->rsa, ctx->padding);
+        if (md_len < 0 || md_len != (int)tbslen)
+            return WOLFSSL_FAILURE;
```

Comment:
`wolfSSL_RSA_public_decrypt` should be returning `-1` on error. Please change `wolfSSL_RSA_public_decrypt` so that it returns `WOLFSSL_FATAL_ERROR` on error.

Reply:
Also is there a reason not to use `wolfSSL_RSA_verify_ex` with `WC_HASH_TYPE_NONE` as the `type` variable?

Reply:
> ```wolfSSL_RSA_public_decrypt``` should be returning -1 on error.

Ok fixed.

> Also is there a reason not to use wolfSSL_RSA_verify_ex with WC_HASH_TYPE_NONE as the type variable?

Because we could not specify hash algorithm for that function. I tried this code but it causes ```This NID (md type) not configured or not implemented```.

```
    case EVP_PKEY_RSA:
        return wolfSSL_RSA_verify_ex(WC_HASH_TYPE_NONE, tbs,
            (unsigned int)tbslen, sig, (unsigned int)siglen, ctx->pkey->rsa,
            ctx->padding);
```

Reply:
>Ok fixed.

Thanks.

>Because we could not specify hash algorithm for that function. I tried this code but it causes This NID (md type) not configured or not implemented.

The reason is that `wolfSSL_EVP_PKEY_sign` is doing a direct `wolfSSL_RSA_private_encrypt` call on the input. This is incorrect as it doesn't produce an actual padded signature. The following patch gets it working correctly. Please apply it in this pull request:
```
julek@julek-MS-7A38:~/Projects/wolfssl-julek$ git diff
diff --git a/wolfcrypt/src/evp.c b/wolfcrypt/src/evp.c
index 10f0b957e3..461fe964e3 100644
--- a/wolfcrypt/src/evp.c
+++ b/wolfcrypt/src/evp.c
@@ -1896,8 +1896,6 @@ int wolfSSL_EVP_PKEY_sign_init(WOLFSSL_EVP_PKEY_CTX *ctx)
 int wolfSSL_EVP_PKEY_sign(WOLFSSL_EVP_PKEY_CTX *ctx, unsigned char *sig,
                         size_t *siglen, const unsigned char *tbs, size_t tbslen)
 {
-    int len = 0;
-
     WOLFSSL_MSG("wolfSSL_EVP_PKEY_sign");
 
     if (!ctx || ctx->op != EVP_PKEY_OP_SIGN || !ctx->pkey)
@@ -1907,19 +1905,22 @@ int wolfSSL_EVP_PKEY_sign(WOLFSSL_EVP_PKEY_CTX *ctx, unsigned char *sig,
     (void)siglen;
     (void)tbs;
     (void)tbslen;
-    (void)len;
 
     switch (ctx->pkey->type) {
 #if !defined(NO_RSA) && !defined(HAVE_USER_RSA)
-    case EVP_PKEY_RSA:
-        len = wolfSSL_RSA_private_encrypt((int)tbslen, tbs, sig,
-              ctx->pkey->rsa, ctx->padding);
-        if (len < 0)
-            break;
-        else {
-            *siglen = len;
-            return WOLFSSL_SUCCESS;
+    case EVP_PKEY_RSA: {
+        unsigned int sigLen;
+        if (siglen == NULL)
+            return WOLFSSL_FAILURE;
+        sigLen = (unsigned int)*siglen;
+        if (wolfSSL_RSA_sign_generic_padding(WC_HASH_TYPE_NONE, tbs, tbslen,
+                sig, &sigLen, ctx->pkey->rsa, 1, ctx->padding)
+                    != WOLFSSL_SUCCESS) {
+            return WOLFSSL_FAILURE;
         }
+        *siglen = (size_t)sigLen;
+        return WOLFSSL_SUCCESS;
+    }
 #endif /* NO_RSA */
 
 #ifndef NO_DSA
@@ -2024,13 +2025,9 @@ int wolfSSL_EVP_PKEY_verify(WOLFSSL_EVP_PKEY_CTX *ctx, const unsigned char *sig,
     switch (ctx->pkey->type) {
 #if !defined(NO_RSA) && !defined(HAVE_USER_RSA)
     case EVP_PKEY_RSA: {
-        unsigned char md[WC_MAX_DIGEST_SIZE];
-        int md_len = wolfSSL_RSA_public_decrypt(
-            (int)siglen, sig, md, ctx->pkey->rsa, ctx->padding);
-        if (md_len < 0 || md_len != (int)tbslen)
-            return WOLFSSL_FAILURE;
-        return XMEMCMP(tbs, md, tbslen) == 0 ?
-            WOLFSSL_SUCCESS : WOLFSSL_FAILURE;
+        return wolfSSL_RSA_verify_ex(WC_HASH_TYPE_NONE, tbs,
+            (unsigned int)tbslen, sig, (unsigned int)siglen, ctx->pkey->rsa,
+            ctx->padding);
     }
 #endif /* NO_RSA */
 
```

Reply:
Thanks, fixed.